moon/monitors.py

```diff
@@ -0,0 +1,94 @@
+"""
+  Moon Monitors for easier strategy coding
+"""
+import math
+
+from osgar.lib.mathex import normalizeAnglePIPI
+
+
+def min_dist(laser_data):
+    if len(laser_data) > 0:
+        # remove ultra near reflections and unlimited values == 0
+        laser_data = [x if x > 10 else 10000 for x in laser_data]
+        return min(laser_data)/1000.0
+    return 0
+
+
+class LidarCollisionException(Exception):
+    pass
+
+
+class LidarCollisionMonitor:
+    def __init__(self, robot):
+        self.robot = robot
+
+    def update(self, robot, channel, data):
+        if channel == 'scan':
+            size = len(robot.scan)
+            # measure distance in front of the rover = 180deg of 270deg
+            if min_dist(robot.scan[size//6:-size//6]) < 1.0:
+                raise LidarCollisionException()
+
+    # context manager functions
+    def __enter__(self):
+        self.callback = self.robot.register(self.update)
+        return self
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        self.robot.unregister(self.callback)
```

Comment:
Would it make sense to create a base/mixin class that would predefine the `__enter__` and `__exit__` calls? We have 3 copies of the same code here which is usually the limit when to start thinking about refactoring.

Reply:
Yes, I also had it in mind & I will change it `BaseMonitor` or `MonitorBase` or some other name scheme recommendation? Thanks

Reply:
I'd go with `BaseMonitor`.